java/clients/src/main/java/sleeper/clients/QueryWebSocketClient.java

```diff
@@ -138,95 +180,184 @@ private Map<String, String> getAwsIamAuthHeaders(URI serverUri) throws URISyntax
             return request.getHeaders();
         }
 
-        public boolean isQueryComplete() {
-            return queryComplete;
+        public boolean hasQueryFinished() {
+            return messageHandler.hasQueryFinished();
         }
 
         public long getTotalRecordsReturned() {
-            return totalRecordsReturned;
+            return messageHandler.getTotalRecordsReturned();
         }
 
         @Override
         public void onOpen(ServerHandshake handshake) {
-            System.out.println("Connected to WebSocket API");
+            messageHandler.onOpen(query, this::send);
+        }
+
+        @Override
+        public void onMessage(String json) {
+            messageHandler.onMessage(json);
+        }
 
-            String queryJson = this.querySerDe.toJson(this.query);
-            System.out.println("Submitting Query: " + queryJson);
-            this.send(queryJson);
-            this.outstandingQueries.add(this.query.getQueryId());
+        @Override
+        public void onClose(int code, String reason, boolean remote) {
+            messageHandler.onClose(reason);
         }
 
         @Override
+        public void onError(Exception error) {
+            messageHandler.onError(error);
+        }
+    }
+
+    public static class WebSocketMessageHandler {
+        private final Gson serde = new GsonBuilder().create();
+        private final Set<String> outstandingQueries = new HashSet<>();
+        private final Map<String, List<String>> records = new TreeMap<>();
+        private final QuerySerDe querySerDe;
+        private final ConsoleOutput out;
+        private boolean queryComplete = false;
+        private boolean queryFailed = false;
+        private long totalRecordsReturned = 0L;
+
+        public WebSocketMessageHandler(QuerySerDe querySerDe, ConsoleOutput out) {
+            this.querySerDe = querySerDe;
+            this.out = out;
+        }
+
+        public void onOpen(Query query, Consumer<String> messageSender) {
+            out.println("Connected to WebSocket API");
+            sendQuery(query, messageSender);
+        }
+
+        private void sendQuery(Query query, Consumer<String> messageSender) {
```

Comment:
It looks like onOpen is the only place this is called, and all that method does is call this. Maybe they should be one method?

Reply:
Thanks, i've combined this method with onOpen